core/src/events/mod.rs

```diff
@@ -0,0 +1,247 @@
+
+        EventMetadataDetails { pallet, variant }
+    }
+
+    /// Return _all_ of the bytes representing this event, which include, in order:
+    /// - The phase.
+    /// - Pallet and event index.
+    /// - Event fields.
+    /// - Event Topics.
+    pub fn bytes(&self) -> &[u8] {
+        &self.all_bytes[self.start_idx..self.end_idx]
+    }
+
+    /// Return the bytes representing the fields stored in this event.
+    pub fn field_bytes(&self) -> &[u8] {
+        &self.all_bytes[self.event_fields_start_idx..self.event_fields_end_idx]
+    }
+
+    /// Decode and provide the event fields back in the form of a [`scale_value::Composite`]
+    /// type which represents the named or unnamed fields that were present in the event.
+    pub fn field_values(&self) -> Result<scale_value::Composite<u32>, Error> {
+        let bytes = &mut self.field_bytes();
+        let event_metadata = self.event_metadata();
+
+        let mut fields = event_metadata
+            .variant
+            .fields
+            .iter()
+            .map(|f| scale_decode::Field::new(&f.ty.id, f.name.as_deref()));
+
+        let decoded =
+            scale_value::scale::decode_as_fields(bytes, &mut fields, self.metadata.types())?;
+
+        Ok(decoded)
+    }
+
+    /// Attempt to decode these [`EventDetails`] into a type representing the event fields.
+    /// Such types are exposed in the codegen as `pallet_name::events::EventName` types.
+    pub fn as_event<E: StaticEvent>(&self) -> Result<Option<E>, Error> {
+        let ev_metadata = self.event_metadata();
+        if ev_metadata.pallet.name() == E::PALLET && ev_metadata.variant.name == E::EVENT {
+            let mut fields = ev_metadata
+                .variant
+                .fields
+                .iter()
+                .map(|f| scale_decode::Field::new(&f.ty.id, f.name.as_deref()));
+            let decoded =
+                E::decode_as_fields(&mut self.field_bytes(), &mut fields, self.metadata.types())?;
+            Ok(Some(decoded))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Attempt to decode these [`EventDetails`] into a root event type (which includes
+    /// the pallet and event enum variants as well as the event fields). A compatible
+    /// type for this is exposed via static codegen as a root level `Event` type.
+    pub fn as_root_event<E: DecodeAsType>(&self) -> Result<E, Error> {
+        let bytes = &self.all_bytes[self.event_start_idx..self.event_fields_end_idx];
+
+        let decoded = E::decode_as_type(
+            &mut &bytes[..],
+            &self.metadata.outer_enums().event_enum_ty(),
+            self.metadata.types(),
+        )?;
+
+        Ok(decoded)
+    }
+
+    /// Return the topics associated with this event.
+    pub fn topics(&self) -> &[T::Hash] {
+        &self.topics
+    }
+}
+
+/// Details for the given event plucked from the metadata.
+pub struct EventMetadataDetails<'a> {
+    pub pallet: PalletMetadata<'a>,
```

Comment:
sorry for being annoying as hell but again maybe add getter for these instead of public fields?